### PR TITLE
Faster Inbound Pipeline

### LINKED ISSUE #79718: [CI] SnapshotStressTestsIT testRandomActivities failing
Test failed because of a MasterNotDiscoveredException. Not sure if this was a random issue or worth investigating so raising the issue.

**Build scan:**
https://gradle-enterprise.elastic.co/s/6wiigan6zqcyy/tests/:server:internalClusterTest/org.elasticsearch.snapshots.SnapshotStressTestsIT/testRandomActivities

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SnapshotStressTestsIT.testRandomActivities" -Dtests.seed=FF84BCB3D914AD7C -Dtests.locale=es-US -Dtests.timezone=America/Dawson -Druntime.java=11`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.snapshots.SnapshotStressTestsIT&tests.test=testRandomActivities

**Failure excerpt:**
```
org.elasticsearch.discovery.MasterNotDiscoveredException: (No message provided)

  at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.onTimeout(TransportMasterNodeAction.java:253)
  at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:324)
  at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:241)
  at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:621)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(Thread.java:834)

```

### PR DESCRIPTION
**Pull request by original-brownbear**
Optimize away a number of refcounting in the inbound pipeline and remove the indirection of collecting pieces of a message into an `ArrayList` before passing them to the aggregator which does its own collecting of these pieces anyway.
The production impact of this change is expected to be relatively minor, though I think it does save a little indirection and might allow for further optimisations down the line. Also, it makes the code more obviously correct by only incrementing the reference count of buffers that are queued for later use.

I found this while looking into why #79718 breaks on OSX CI while logging an endless stream of slow transport warnings. Theses are in large part caused by instantiating lots and lots of leak tracking exceptions with stack-traces in tests. This change reduces the number of increments and decrements massively, thus speeding up tests (on all platforms not just OSX for what it's worth).

Rough illustration (no real benchmark, just some profiling from tests but you can see the right hand side of the profile going down even for an expensive transport action like get-snapshots with many pending snapshots). I don't have the charts for other tests with me right now, but the effect is obviously much bigger when dealing with smaller transport messages in tests:

before:

image

after:

image

closes #79718 (I assume this will be enough to stabilize the test now combined with the other fix already applied to the issue)

### CODE REVIEW

**Comment by original-brownbear** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 1]
```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
...
-                        if (fragment instanceof ReleasableBytesReference) {
-                            ((ReleasableBytesReference) fragment).close();
-                        }
-                    }
-                    fragments.clear();
-                }
+                break;
             }
         }
+        // if handling the messages didn't cause the channel to get closed and we did not fully consume the buffer retain it
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:109]: Could we have a test in `InboundPipelineTests` showing that we handle the `isClosed` case correctly here? And elsewhere I guess, but I checked that we never actually exercise the `isClosed == true` branch here.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 2]
```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
...
+                    bytesDecoded = decode(channel, toDecode);
                 }
+            }
+            if (bytesDecoded != 0 && isClosed == false) {
+                releasePendingBytes(bytesDecoded);
             } else {
-                assert aggregator.isAggregating();
-                assert fragment instanceof ReleasableBytesReference;
-                aggregator.aggregate((ReleasableBytesReference) fragment);
+                return;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:132]: If `isClosed` can we assert that `pending` is now empty?

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 3]
```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
...
-                    messageHandler.accept(channel, aggregated);
+    private int decode(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
+        return decoder.decode(channel, reference, this::forwardFragment);
+    }
+
+    private void doHandleBytesWithPending(TcpChannel channel) throws IOException {
+        do {
+            final int bytesDecoded;
+            if (pending.size() == 1) {
+                bytesDecoded = decode(channel, pending.peekFirst());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:123]: If the channel is closed then the bytes get released mid-decode. Is that a problem? (possibly not, but I'd like to hear the reasoning)

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 4]
```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
...
+            if (bytesDecoded != 0 && isClosed == false) {
+                releasePendingBytes(bytesDecoded);
             } else {
-                assert aggregator.isAggregating();
-                assert fragment instanceof ReleasableBytesReference;
-                aggregator.aggregate((ReleasableBytesReference) fragment);
+                return;
             }
-        }
+        } while (isClosed == false && pending.isEmpty() == false);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:134]: If `isClosed` didn't we already return?

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java [hunk 5]
```diff
@@ -38,23 +38,31 @@ public InboundDecoder(Version version, Recycler<BytesRef> recycler) {
         this.recycler = recycler;
     }
 
-    public int decode(ReleasableBytesReference reference, Consumer<Object> fragmentConsumer) throws IOException {
+    public int decode(
+        TcpChannel channel,
+        ReleasableBytesReference reference,
+        CheckedBiConsumer<TcpChannel, Object, IOException> fragmentConsumer
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:44]: Could we not pass the `TcpChannel` around everywhere like this? It's always the same channel isn't it?

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java [hunk 6]
```diff
@@ -19,15 +19,13 @@
...
 import java.util.ArrayDeque;
-import java.util.ArrayList;
 import java.util.function.BiConsumer;
 import java.util.function.Function;
 import java.util.function.LongSupplier;
 import java.util.function.Supplier;
 
 public class InboundPipeline implements Releasable {
 
-    private static final ThreadLocal<ArrayList<Object>> fragmentList = ThreadLocal.withInitial(ArrayList::new);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:30]: :+1:

**Review by DaveCTurner — COMMENTED** [review]
I'm not a big fan of this change, I'd prefer to keep the ref counting more obviously correct and deal with the leak-detector test slowness directly: perhaps stop collecting stack traces on the platforms on which it's too slow to cope while we're not chasing any particular leak there, or else just extend timeouts as needed. I left a few inline comments too.

**Comment by original-brownbear** [comment]
> I'd prefer to keep the ref counting more obviously correct 

IMO this is exactly what this PR does. We shouldn't be ref-count incrementing and decrementing left and right when we don't need it. We should increment when we need to hold on to something outside of the current execution stack and only then, anything else just makes the code needlessly hard to follow IMO. To me incrementing a reference always means we fork off or need something later and this wasn't the case here at all.

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:44]: Probably, I bet we can make additional optimizations like that (though I might actually like it better to be less "functional" here and just remove the ability to pass a consumer here when we only have one production code user of this method). It's weird to have this kind of generality on the very very hot path (though admittedly it's not showing up in profiling) just for tests?

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:123]: The channel is closed on the same thread that this runs on, we don't release mid-decode. We release after deserializing and handling what we deserialized in all cases.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:109]: Let me try :) Should be doable, these paths are exercised by some internal cluster tests so there should be an obvious way to do it. On it

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 4 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:134]: ++ right

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:132]: Yea that should be fine, will add

**Comment by DaveCTurner** [comment]
I sort of see what you mean but also this change means that refcounting is different depending on whether there are multiple fragments to combine or not. I don't see a way to avoid the extra refcounting in the multi-fragment case, but it isn't as simple as it was and it took some thought.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:123]: Yeah I really just meant within the `decode()` method.

**Comment by tbrooks8** [comment]
I looked at this briefly yesterday. I was also opposed to the change. I don’t think it fixes the underlying issue. If normal ref count incs and decs kill the test framework, I think that is an issue. And removing incs/decs used in production does not seem like a good solution. 

Additionally this code was structured similar to Netty specifically to separate decoding from aggregation, from handling. I didn’t want the decoder wrapping further pipeline steps. Always attempted to step back to the top (Pipeline) before moving to the next step. I did not want the decoder state to be concerned about what happened further down the pipeline.

The Consumer interface was just for testing. If I did it again I would probably pass the list in similar to netty.  

I also don’t agree on the correctness point. try with resource for each scope and passing a new retain into another scope guarantees an uncaught exception does not lead to a leak. 

I’m not rejecting the PR if anything. You all can find whatever approach you want. Those were just my thoughts looking at it.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:123]: Ah I see. I think there we're good because we don't have any incremental decode on a message. We decode a full message and pass it along to message handling, then return from decode. So closing the channel will always be the last step in `decode` as far as I can see.

**Comment by original-brownbear** [comment]
>  perhaps stop collecting stack traces on the platforms on which it's too slow to cope while we're not chasing any particular leak there

One important point I'd make here is that the problem is not necessarily just that we collect stack traces for every message. It's simply how many we have to collect with the redundant ref counting we do. This is the traces we get for a tiny message (so no aggregation or anything in here) once it hits `InboundHandler`, so before we even get around to deserializing in master.

```

Recent access records: 
#1:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retain(ReleasableBytesReference.java:81)
	org.elasticsearch.transport.InboundAggregator.aggregate(InboundAggregator.java:80)
	org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:153)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:120)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#2:
	org.elasticsearch.common.bytes.ReleasableBytesReference.close(ReleasableBytesReference.java:96)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:113)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#3:
	org.elasticsearch.common.bytes.ReleasableBytesReference.close(ReleasableBytesReference.java:96)
	org.elasticsearch.transport.InboundPipeline.releasePendingBytes(InboundPipeline.java:188)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:106)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#4:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retain(ReleasableBytesReference.java:81)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retainedSlice(ReleasableBytesReference.java:87)
	org.elasticsearch.transport.InboundDecoder.internalDecode(InboundDecoder.java:94)
	org.elasticsearch.transport.InboundDecoder.decode(InboundDecoder.java:44)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:104)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#5:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retain(ReleasableBytesReference.java:81)
	org.elasticsearch.transport.InboundPipeline.getPendingBytes(InboundPipeline.java:164)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:103)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#6:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retainedSlice(ReleasableBytesReference.java:90)
	org.elasticsearch.transport.InboundPipeline.releasePendingBytes(InboundPipeline.java:183)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:106)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
#7:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retain(ReleasableBytesReference.java:81)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:95)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:85)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
Created at:
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:347)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
: 3 leak records were discarded because they were duplicates
```

After my change it's this:

```

Recent access records: 
#1:
	org.elasticsearch.core.AbstractRefCounted.incRef(AbstractRefCounted.java:24)
	org.elasticsearch.common.bytes.ReleasableBytesReference.retain(ReleasableBytesReference.java:81)
	org.elasticsearch.transport.InboundAggregator.aggregate(InboundAggregator.java:80)
	org.elasticsearch.transport.InboundPipeline.forwardFragment(InboundPipeline.java:157)
	org.elasticsearch.transport.InboundDecoder.internalDecode(InboundDecoder.java:124)
	org.elasticsearch.transport.InboundDecoder.decode(InboundDecoder.java:48)
	org.elasticsearch.transport.InboundPipeline.decode(InboundPipeline.java:116)
	org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:102)
	org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:83)
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:351)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
Created at:
	org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:347)
	org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:222)
	org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:35)
	org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:128)
	org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:143)
	org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:415)
	org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:241)
	org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:168)
	org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:125)
	java.base/java.lang.Thread.run(Thread.java:833)
```

... IMO this is logically much closer to the reality of what gets retained where and makes it far easier to understand where a leak may be coming from when debugging.

I agree that many of the redundant ref-count changes we have right now are trivial to reason about because they're all try-with-resources, but they also don't add anything because of their simplicity. They burn cycles now (though not that many in production) and make debugging a leak needlessly hard, because you have to go through all of them and make sure none of the 7 (instead of 2 that are needed) recorded touch points are correct.

Also. without this change we are not actually operating the way Netty operates either I'd say because we'd just deserialize messages one by one anyway and then pass them along the pipeline as soon as we had one full message instead of Netty which at least optionally batches things in decoding here and there via the flag `io.netty.handler.codec.ByteToMessageDecoder#singleDecode`? Since we don't have that kind of batching I don't really see the point in pushing stuff up the stack again like that, it just adds complexity for working with the list that is always <=3 in length anyway?

**Comment by original-brownbear** [comment]
>  I don't see a way to avoid the extra refcounting in the multi-fragment case, but it isn't as simple as it was and it took some thought.

True, but that simplicity also meant a lot of needless cycles went into wrapping trivial messages multiple times. In a world where we cache 3-element lists thread-local because we think this code is hot enough to justify it, I find this to be an acceptable bit of complexity to add.

**Comment by DaveCTurner** [comment]
Yeah it's neater for sure, but that's still fundamentally a bug in the test-only leak tracker for which we're proposing a production code change. If we properly paired every acquire with its corresponding release then we could clean up the stack traces for the acquires that weren't released.

Have we considered not separating the parts of the `InboundPipeline` so much? Acknowledging that Tim says we're following Netty's patterns here, but Netty is super-flexible so it needs these things to be separate whereas we have much simpler needs.

**Comment by original-brownbear** [comment]
>  but Netty is super-flexible so it needs these things to be separate whereas we have much simpler needs.

I'd be all for it. Technically, there is no reason for the code to be spread out like this in our case IMO. Netty allows for things like multiple layers of decoding and aggregating messages where tracking a list makes sense to neatly do stuff like read a chunked HTTP message step by step and whatnot. We don't do anything like that and IMO could just make the code as flat as the ref counting is now, rather than have redundant complicated ref-counting to wrap around all the various steps in the current code.

> but that's still fundamentally a bug in the test-only leak tracker for which we're proposing a production code change. I

Not sure I fully agree with this. I'd say yes, I wouldn't have proposed this without the leak-tracker causing us trouble, but I think it's worthwhile in isolation to not have needless wrapping of buffers and ref counting in code this hot. Also, I still don't see how redundant ref count incrementing and releasing without forking adds anything but confusion to this code (the case where we handle aggregate and single buffer message is the only exception to this where it maybe does).

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:44]: Hiding my above comment, I misread the question initially. I added passing around the channel here to avoid instantiating a capturing lambda here where previously we had a non-capturing one.

**Comment by original-brownbear** [comment]
Build failure is just a Jenkins issue, the build (part-2) went through fine.

File: server/src/main/java/org/elasticsearch/transport/InboundPipeline.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundPipeline.java:109]: Test added, should be good for another round now :)

File: server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java [hunk 7]
```diff
@@ -82,6 +82,20 @@ public ReleasableBytesReference retain() {
         return this;
     }
 
+    public ReleasableBytesReference releasableSlice(int from) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java:85]: This seems to be like a "transfer" ownership operation. It seems to me that it is a bit artificial in that it's primary purpose is to avoid an incref/decref solely for the leak tracker.

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java [hunk 8]
```diff
@@ -84,33 +92,39 @@ public int internalDecode(ReleasableBytesReference reference, Consumer<Object> f
...
 
             int bytesConsumedThisDecode = 0;
             if (decompressor != null) {
-                bytesConsumedThisDecode += decompress(retainedContent);
+                bytesConsumedThisDecode += decompress(
+                    maxBytesToConsume == remainingToConsume ? reference.slice(0, maxBytesToConsume) : reference
+                );
                 bytesConsumed += bytesConsumedThisDecode;
                 ReleasableBytesReference decompressed;
                 while ((decompressed = decompressor.pollDecompressedPage(isDone())) != null) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:108]: I wonder if a push-style interaction out of `decompress` would be easier, since then the ref-counting is mostly not an issue.

**Comment by original-brownbear** [comment]
https://github.com/elastic/elasticsearch/pull/81096 makes this PR a clean win in production IMO. Without this change, Netty's default 4 target records on leak tracking will not be enough to record the full history of a leak. With this change, it should be fine in most cases.

**Review by henningandersen — COMMENTED** [review]
I am also hesitant on this change as is. Left a couple of comments. But I think it could make sense to collapse `InboundDecoder` and `InboundPipeline` as a (separate) experiment to see if this simplifies things

File: server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java (hunk 7 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java:85]: I found the whole thing rather artificial before. We would retain the original, then slice and then build a new releasable from the original. This way at least, we're honest about the fact that this is all backed by the same ref-count and all we really did is move the read offset. In the end part of the motivation for this change is that we aren't actually transferring ownership all the time in the two spots that this method is used in.

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 8 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:108]: True, I guess that would be another possible simplification and would make decompress work more similarly to how the rest of the pipeline works with this change as well. Not sure we'd want to blow this one up with making that change here also. I'd rather push that down to a follow-up to keep this somewhat tricky change small? :)

**Comment by original-brownbear** [comment]
> . But I think it could make sense to collapse InboundDecoder and InboundPipeline as a (separate) experiment to see if this simplifies things

++ 

> I am also hesitant on this change as is. 

Did my best to respond to inline comments :) But I'm not really sure what I can do here to address general concerns I'm afraid. The remove ref counting didn't actually add anything safety (with the one exception mentioned). Yes it was safe, but it was also redundant and obscured the points where actual transfers of ownership took place?

**Comment by tbrooks8** [comment]
1. It seems very plausible that we could introduce some type of chunked content scheme in the near future. As we move to a world were compression is enabled by default, it is unfortunate that a message must be compressed in a single go. That is a world were our decoding step might be more complicated and produce multiple things before an entire message is available.
2. I am very opposed to collapsing the decoder into the pipeline. the decoder is separate specifically because unit testing this code is very important.
3. The `incs` and `decs` around `InboundPipeline#getPendingBytes()` and `InboundPipeline#releasePendingBytes` seem fine to remove. Same thing with the retains in `InboundDecoder#internalDecode` before we pass to the decompressor.
4. I continue to opposed calling further steps down the pipeline from inside the `InboundDecoder`.
5. I don't think any of passing bytes back and forth here is a hot enough code path where uncontented volatile read/writes matter. Particularly since all of this code is surrounded by TLS and Compression which is many many many times more expensive.
6. Compression is poll based as the method must return the number of bytes consumed and can produce multiple pages. I think we do not want to interrupt a decompression loop to pass bytes down the pipeline.

**Comment by original-brownbear** [comment]
Pinging again on this one. I'm still of the opinion that this is an outright win because of the fact that it fixes leak-tracking to be more useful in the real world (https://github.com/elastic/elasticsearch/pull/80656#issuecomment-979411349). Can we do anything to push this forward or simplify our buffer ref-counting to more closely reflect what gets escaped where? 
My main motivation here at this point is fixing the leak tracking.

**Comment by original-brownbear** [comment]
closing in favor of https://github.com/elastic/elasticsearch/pull/123390
